### PR TITLE
Implement profile serialization

### PR DESCRIPTION
**Pull request by chenkins**

### CODE REVIEW

**Review by dkocher — APPROVED** [review]

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 1]
```diff
@@ -39,11 +39,137 @@
...
+            dict.setStringForKey(this.getOAuthTokenUrl(), OAUTH_TOKEN_URL_KEY);
+        }
+        if(this.getOAuthRedirectUrl() != null) {
+            dict.setStringForKey(this.getOAuthRedirectUrl(), OAUTH_REDIRECT_URL_KEY);
+        }
+        if(this.getOAuthAuthorizationUrl() != null) {
+            dict.setStringForKey(this.getOAuthAuthorizationUrl(), OAUTH_AUTHORIZATION_URL_KEY);
+        }
+        if(!this.isOAuthPKCE()) {
+            dict.setStringForKey(Boolean.toString(this.isOAuthPKCE()), OAUTH_PKCE_KEY);
```
Comment (chenkins) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:64]: @dkocher `setBooleanForKey` here as well?

Comment (dkocher) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:64]: Also all `*_CONFIGURABLE_KEY*.

**Review by dkocher — CHANGES_REQUESTED** [review]
Write boolean instead of string value for all occurrences of `Boolean.toString`.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java (hunk 1 above)
Comment (chenkins) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java]: Done.

**Review by dkocher — CHANGES_REQUESTED** [review]
Squash commits.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 2]
```diff
@@ -178,11 +304,19 @@ public String getHostnamePlaceholder() {
 
     @Override
     public String getUsernamePlaceholder() {
+        return getDefaultUsernamePlaceholder();
+    }
+
+    private static String getDefaultUsernamePlaceholder() {
```
Comment (dkocher) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:310]: Why is this required?

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 3]
```diff
@@ -178,11 +304,19 @@ public String getHostnamePlaceholder() {
...
         return LocaleFactory.localizedString("Username", "Credentials");
     }
 
     @Override
     public String getPasswordPlaceholder() {
+        return getDefaultPasswordPlaceholder();
+    }
+
+    private static String getDefaultPasswordPlaceholder() {
```
Comment (dkocher) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:319]: Why is this required?

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java (hunk 2 above)
Comment (chenkins) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:310]: No need to match duplicated `LocaleFactory.localizedString("Username", "Credentials");` in multiple places.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java (hunk 3 above)
Comment (chenkins) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:319]: Dito.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java (hunk 2 above)
Comment (dkocher) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:310]: There is a single occurrence only and such duplicated code does no harm.

**Review by dkocher — APPROVED** [review]

File: core/src/test/resources/AbstractProtocolTest/full.cyberduckprofile [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	<key>Anonymous Configurable</key>
+	<true/>
+	<key>Token Configurable</key>
+	<true/>
+	<key>Certificate Configurable</key>
+	<true/>
+	<key>Private Key Configurable</key>
+	<true/>
+	<key>Properties</key>
+	<dict>
```
Comment (chenkins) [inline core/src/test/resources/AbstractProtocolTest/full.cyberduckprofile:91]: @dkocher shouldn't this be an array as in https://github.com/search?q=repo%3Aiterate-ch%2Fprofiles%20Properties&type=code

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 5]
```diff
@@ -39,11 +39,137 @@
...
+            dict.setStringForKey(this.getSTSEndpoint(), STS_ENDPOINT_KEY);
+        }
+        if(this.disk() != null) {
+            dict.setStringForKey(this.disk(), DISK_KEY);
+        }
+        if(this.icon() != null) {
+            dict.setStringForKey(this.icon(), ICON_KEY);
+        }
+        if(this.getIdentifier() != null) {
+            dict.setStringForKey(this.getIdentifier(), PROTOCOL_KEY);
```
Comment (chenkins) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:79]: @dkocher bad code smell, we implement `AbstractHostProtocol.serialize` but want to (mis-)use at as a builder for profile serialization?

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 6]
```diff
@@ -39,11 +39,137 @@
...
+        }
+        if(this.getProvider() != null) {
+            dict.setStringForKey(this.getProvider(), VENDOR_KEY);
```
Comment (chenkins) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:82]: dito.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java [hunk 7]
```diff
@@ -39,11 +39,137 @@
...
+            dict.setBooleanForKey(this.isOAuthConfigurable(), OAUTH_CONFIGURABLE_KEY);
+        }
+        if(this.isCertificateConfigurable()) {
+            dict.setBooleanForKey(this.isCertificateConfigurable(), CERTIFICATE_CONFIGURABLE_KEY);
+        }
+        if(this.isPrivateKeyConfigurable()) {
+            dict.setBooleanForKey(this.isPrivateKeyConfigurable(), PRIVATE_KEY_CONFIGURABLE_KEY);
+        }
+        if(!this.getProperties().isEmpty()) {
+            dict.setMapForKey(this.getProperties(), PROPERTIES_KEY);
```
Comment (chenkins) [inline core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:164]:
```suggestion
            dict.setStringListForKey(this.getProperties().entrySet().stream().map(entry -> String.format("%s=%s", entry.getKey(), entry.getValue())).collect(Collectors.toList()), PROPERTIES_KEY);
```

File: core/src/test/resources/AbstractProtocolTest/full.cyberduckprofile [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		<key>My Key</key>
+		<string>My Value</string>
+	</dict>
```
Comment (chenkins) [inline core/src/test/resources/AbstractProtocolTest/full.cyberduckprofile:94]:
```suggestion
	<array>
		<string>My Key=My Value</string>
	</array>
```
```suggestion
	<dict>
		<key>My Key</key>
		<string>My Value</string>
	</dict>
```

Comment (chenkins) [inline core/src/test/resources/AbstractProtocolTest/full.cyberduckprofile:94]:
```suggestion
	<array>
		<string>My Key=My Value</string>
	</array>
```

**Review by ylangisc — COMMENTED** [review]
@dkocher There are unresolved conversations.

File: core/src/main/java/ch/cyberduck/core/AbstractProtocol.java (hunk 5 above)
Comment (dkocher) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:79]: I do not understand the concern.

Comment (chenkins) [reply core/src/main/java/ch/cyberduck/core/AbstractProtocol.java:79]: `VENDOR_KEY`/`PROTOCOL_KEY` is defined in `Profile` (as it should), as `AbstractProtocol` is not a sub-class of `Profile`, it should have no notion of `VENDOR_KEY`/`PROTOCOL_KEY`. It works but is confusing imho 😕

File: core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java [hunk 9]
```diff
@@ -270,4 +271,29 @@ public boolean isEnabled() {
...
+                return false;
+            }
+
+            @Override
+            public String getIdentifier() {
+                return "MyId";
+            }
+        })));
+        final Profile profile = reader.read(
+                new Local("src/test/resources/full.cyberduckprofile")
```
Comment (chenkins) [inline core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java:294]: Use `Paths.get(resource.toURI()).toFile().getAbsolutePath()` in order to not depend on current working directory.

File: core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java [hunk 10]
```diff
@@ -270,4 +271,29 @@ public boolean isEnabled() {
...
+        );
+        final String expected = IOUtils.toString(this.getClass().getResource("/full.cyberduckprofile"));
```
Comment (chenkins) [inline core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java:296]: Should we test output of `PlistWriter.write()` against `expected` as well?

File: core/src/main/java/ch/cyberduck/core/Profile.java [hunk 11]
```diff
@@ -117,7 +117,33 @@ public Profile(final Protocol parent, final Deserializer<?> dict) {
...
+                case PORT_CONFIGURABLE_KEY:
+                case PATH_CONFIGURABLE_KEY:
+                case USERNAME_CONFIGURABLE_KEY:
+                case PASSWORD_CONFIGURABLE_KEY:
+                case ANONYMOUS_CONFIGURABLE_KEY:
+                case TOKEN_CONFIGURABLE_KEY:
+                case OAUTH_CONFIGURABLE_KEY:
+                case CERTIFICATE_CONFIGURABLE_KEY:
+                case PRIVATE_KEY_CONFIGURABLE_KEY:
+
```
Comment (chenkins) [inline core/src/main/java/ch/cyberduck/core/Profile.java:131]: empty line

File: core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java (hunk 10 above)
Comment (dkocher) [reply core/src/test/java/ch/cyberduck/core/serializer/impl/dd/ProfilePlistReaderTest.java:296]: Implementation is same as `serialize(new PlistSerializer()).toXMLPropertyList()`
